Code/Framework/AzCore/AzCore/DOM/DomValue.cpp

```diff
@@ -0,0 +1,1305 @@
+        array.push_back(AZStd::move(value));
+        return *this;
+    }
+
+    Value& Value::PopBack()
+    {
+        GetArrayInternal().pop_back();
+        return *this;
+    }
+
+    Array::Iterator Value::Erase(Array::ConstIterator pos)
+    {
+        return GetArrayInternal().erase(pos);
+    }
+
+    Array::Iterator Value::Erase(Array::ConstIterator first, Array::ConstIterator last)
+    {
+        return GetArrayInternal().erase(first, last);
+    }
+
+    Array::ContainerType& Value::GetMutableArray()
+    {
+        return GetArrayInternal();
+    }
+
+    const Array::ContainerType& Value::GetArray() const
+    {
+        return GetArrayInternal();
+    }
+
+    void Value::SetNode(AZ::Name name)
+    {
+        m_value = AZStd::allocate_shared<Node>(StdValueAllocator(), AZStd::move(name));
+    }
+
+    void Value::SetNode(AZStd::string_view name)
+    {
+        SetNode(AZ::Name(name));
+    }
+
+    AZ::Name Value::GetNodeName() const
+    {
+        return GetNodeInternal().GetName();
+    }
+
+    void Value::SetNodeName(AZ::Name name)
+    {
+        GetNodeInternal().SetName(AZStd::move(name));
+    }
+
+    void Value::SetNodeName(AZStd::string_view name)
+    {
+        SetNodeName(AZ::Name(name));
+    }
+
+    void Value::SetNodeValue(Value value)
+    {
+        AZ_Assert(GetType() == Type::Node, "AZ::Dom::Value: Attempted to set value for non-node type");
+        Array::ContainerType& nodeChildren = GetArrayInternal();
+
+        // Set the first non-node child, if one is found
+        for (Value& entry : nodeChildren)
+        {
+            if (entry.GetType() != Type::Node)
+            {
+                entry = AZStd::move(value);
+                return;
+            }
+        }
+
+        // Otherwise, append the value entry
+        nodeChildren.push_back(AZStd::move(value));
+    }
+
+    Value Value::GetNodeValue() const
+    {
+        AZ_Assert(GetType() == Type::Node, "AZ::Dom::Value: Attempted to get value for non-node type");
+        const Array::ContainerType& nodeChildren = GetArrayInternal();
+
+        // Get the first non-node child, if one is found
+        for (const Value& entry : nodeChildren)
+        {
+            if (entry.GetType() != Type::Node)
+            {
+                return entry;
+            }
+        }
+
+        return Value();
+    }
+
+    Node& Value::GetMutableNode()
+    {
+        return GetNodeInternal();
+    }
+
+    const Node& Value::GetNode() const
+    {
+        return GetNodeInternal();
+    }
+
+    int64_t Value::GetInt64() const
+    {
+        switch (m_value.index())
+        {
+        case GetTypeIndex<int64_t>():
+            return AZStd::get<int64_t>(m_value);
+        case GetTypeIndex<uint64_t>():
+            return aznumeric_cast<int64_t>(AZStd::get<uint64_t>(m_value));
+        case GetTypeIndex<double>():
+            return aznumeric_cast<int64_t>(AZStd::get<double>(m_value));
```

Comment:
I'm going back and forth on this - in rapidjson there's explicitly just a `numeric` type and the internal storage type is made completely opaque - we're more transparent here, but I can see annoyances here and there when not being able to retrieve / trivially convert numeric types e.g. you serialize a signed int with a value of `5`, the deserializer reads it in as a uint64 since there's no sign, suddenly you can't retrieve the value with `GetInt64()`.

The options I see are:

1. No comparisons or implicit conversions, you have to check types and convert.
2. No implicit comparisons, but conversions happen implicitly (so if you have a uint64 and call `GetInt64()` you'll get the `aznumeric_cast`ed value).
3. Continue to do implicit conversions and comparisons (current route) - anything more esoteric like bool -> int still requires checking by hand
4. More conversions! "5.5" can be parsed into a double, do that on the value side on-demand, etc.

I think 4 is definitely a serializer thing, but I like at least being able to retrieve / convert types easily, so I'm leaning towards 2 or 3, with a slight bias towards 3 as the current behavior, but I'm open to more argument for 1, it just might lead to some ergonomic issues.

Reply:
Went with option 2 after offline discussion, getters and setters for other types were also removed (but constructors remained, so override resolution for `int32_t` etc still work when instantiating a `Value`).